### ISSUE TITLE
Gradient on message-passing layers tiny, but fine elsewhere.

### ISSUE THREAD

**Issue by Coprosmo** [rusty1s/pytorch_geometric#1963]
## ❓ Questions & Help

**Some Background:**
I have attempted to reconstruct the GNN described in [this paper](https://arxiv.org/pdf/1905.10006.pdf) to perform graph-level classification on a binary tree-type graph. The left and right children of a node are distinguished by differing edge labels (0 and 1). All node and edge features are fed through multi-layer perceptrons to generate size 128 feature vectors.

The message-passing phase then operates as follows:

1. For a node `n' Messages are generated by concatenating the neighbor's embedding with n's embedding, and the embedding of the edge connecting them. Depending on whether the neighbor is a child or parent of n, this concatenation is passed through one of two different MLPs (to distinguish children and parents).
2. n's embedding is concatenated with the sum of the child embeddings (divided by the number of children), and the sum of the parent embeddings (also divided).
3. Finally, this concatenation is passed through another MLP, and added to n's embedding to form n's new embedding.

To generate the final graph-level embedding, the feature dimension is upsampled to size 1024, and a max-pool is taken. 

**My Issue**
Many of the layers in my neural network show little change throughout training. Initially I thought this might be a case of the vanishing gradient problem, except for the fact that the first layer (which generates the initial node embeddings) shows significant change. The message-passing layers; however, barely change at all.

I discovered that the gradients on the message-passing layers are very small, when I examined the Weights & Biases parameter and gradient view:

See that for this layer, the gradients are small, and the parameters barely change throughout training:
![image](https://user-images.githubusercontent.com/37574545/103467856-3ac14d00-4db8-11eb-8b42-b89f8abd7fd1.png)

Compared to this layer (in the initial node embedding MLP, which is before the above layer) which shows significant change. Note that the gradients are an order of magnitude higher:
![image](https://user-images.githubusercontent.com/37574545/103467490-4b6fc400-4db4-11eb-829c-c3c6278aa5c2.png)

However, there were a couple of layers in the message-passing section which did show to be learning (note the high gradients):
![image](https://user-images.githubusercontent.com/37574545/103467423-cf757c00-4db3-11eb-8e7e-b070568297cc.png)

I can't for the life of me figure out what might be causing this, other than that my code might not be doing what I think it is.

**My Code**

Note that my code differs slightly outside of the message-passing scheme. In particular, to generate the graph-level embedding I take the max-pool and avg-pool over the feature dimension, concatenate these, and pass them through a MLP.

```Python
class MLP(torch.nn.Module):
    """
    Standard MLP scheme for use in Subgraph Pooling model
    """
    def __init__(self, in_channels, out_channels, dim=0):
        super(MLP, self).__init__()
        self.lin1 = Linear(in_channels, 256, dim)
        self.hidden = Linear(256, 128)
        self.lin2 = Linear(128, 128)
        
    def forward(self, x):
        x = F.selu(self.lin1(x))
        x = F.dropout(x, 0.5)
        x = F.selu(self.hidden(x))
        x = F.dropout(x, 0.5)
        x = F.selu(self.lin2(x))
        return x

    
class PaliwalMP(MessagePassing):
    """Define the message-passing scheme from Subgraph Pooling paper."""
    
    def __init__(self, in_channels, out_channels):
        super(PaliwalMP, self).__init__(aggr='mean', flow='source_to_target') #  "Mean" aggregation.
        
        # MLP for Parents and Children, step 2 of Paliwal MP
        self.MLP_edge = BuildingBlock(3*in_channels, in_channels)
        self.MLP_edge_hat = BuildingBlock(3*in_channels, in_channels)
        
        # MLP to pass aggregated message through, step 3 of Paliwal MP
        self.MLP_aggr = BuildingBlock(3*in_channels, in_channels)

        
    def forward(self, x, edge_index_parents, edge_index_children, edge_attr_parents, edge_attr_children):
        # x has shape [N, in_channels]
        # edge_index_x has shape [2, E/2]
        out_parents = self.propagate(edge_index_parents, 
                                     x=x, 
                                     edge_attr=edge_attr_parents, 
                                     direction='up', 
                                     size=None)
        out_children = self.propagate(edge_index_children, 
                                      x=x, 
                                      edge_attr=edge_attr_children, 
                                      direction='down', 
                                      size=None)
        
        out = torch.cat([x, out_parents, out_children], dim=1)
        out = self.MLP_aggr(out) + x
        
        return out

    
    def message(self, x_i, x_j, edge_attr, direction):

        s_ij = torch.cat([x_i, x_j, edge_attr], dim=1)
        if direction == 'up':
            s_ij = self.MLP_edge(s_ij)
        elif direction == 'down':
            s_ij = self.MLP_edge_hat(s_ij)
        
        return s_ij

    
    def update(self, aggr_out, x):
        # aggr_out has shape [N, out_channels]
        return aggr_out


class BuildingBlock(torch.nn.Module):
    """
    Standard MLP scheme for use in Subgraph Pooling model
    """
    def __init__(self, in_channels, out_channels):#, dim=0):
        super(BuildingBlock, self).__init__()
        self.lin1 = Linear(in_channels, 256)
        self.hidden = Linear(256, 128)
        self.lin2 = Linear(128, 128)
        
    def forward(self, x):
        x = F.selu(self.lin1(x))
        x = F.dropout(x, 0.5)
        x = F.selu(self.hidden(x))
        x = F.dropout(x, 0.5)
        x = F.selu(self.lin2(x))
        return x
    

class PaliwalNet(torch.nn.Module):
    """
    Implement GNN from Subgraph Pooling model. Accepts an arbitrary-size graph and produces a scalar output value.
    """
    def __init__(self, t, n_distinct_features, n_classes=4, no_upsample=False, sigmoid=True, softmax=False):
        super(PaliwalNet, self).__init__()
        
        self.embedding = Embedding(num_embeddings=n_distinct_features+1, embedding_dim=128)
        
        self.MLP_V = BuildingBlock(128, 128)
        if t > 0:
            self.MLP_E = BuildingBlock(1, 128)
        
        self.message_passing_steps = nn.ModuleList()
        for i in range(t):
            self.message_passing_steps.append(PaliwalMP(128, 128))

        self.conv1 = nn.Conv1d(128, 512, 1)
        self.conv2 = nn.Conv1d(512, 1024, 1)

        self.lin1 = Linear(2048, 1024)
        self.lin2 = Linear(1024, 512)
        self.lin3 = Linear(512, 256)
        self.lin4 = Linear(256, 128)
        self.lin5 = Linear(128, 128)
        self.lin6 = Linear(128, 4)
        
        self.bn1 = nn.BatchNorm1d(1024)
        self.bn2 = nn.BatchNorm1d(512)
        self.bn3 = nn.BatchNorm1d(256)
        self.bn4 = nn.BatchNorm1d(128)
        self.bn5 = nn.BatchNorm1d(128)
        
        self.softmax = nn.LogSoftmax(dim=0)

  
    def forward(self, data):
        x, edge_index, edge_attr, batch = data.x, data.edge_index, data.edge_attr, data.batch
        edge_index_u, edge_index_d = torch.split(edge_index, int(edge_index.shape[1]/2), dim=1)
        edge_attr_u, edge_attr_d = torch.split(edge_attr, int(edge_attr.shape[0]/2))
        
        # Generate learnable embeddings for node features
        x = x.squeeze(-1)
        x = self.embedding(x)
        x = self.MLP_V(x)
        
        # Embed node and edge features into high dimensional space
        if len(self.message_passing_steps) > 0:
            edge_attr_u = self.MLP_E(edge_attr_u.float())
            edge_attr_d = self.MLP_E(edge_attr_d.float())
        
        for i, message_passing_step in enumerate(self.message_passing_steps):
            x = message_passing_step(x, edge_index_u, edge_index_d, edge_attr_u, edge_attr_d)
        

        # Upsample feature vectors to size 1024 using 1x1 convolution
        x = x.transpose(0,1).unsqueeze(0)
        x = F.selu(self.conv1(x))
        x = F.selu(self.conv2(x))
        x = x.squeeze(0).transpose(0,1)
        
        
        # Final prediction network
        x = torch.cat([gmp(x, batch), gap(x, batch)], dim=1)
        
        x = F.selu(self.lin1(x))
        x = self.bn1(x)
        x = F.selu(self.lin2(x))
        x = self.bn2(x)
        x = F.selu(self.lin3(x))
        x = self.bn3(x)
        x = F.selu(self.lin4(x))
        x = self.bn4(x)
        x = F.selu(self.lin5(x))
        x = self.bn5(x)
        
        x = self.softmax(self.lin6(x))
        
        return x
```

Note: edge_index_u, and edge_index_d represent the upward edges (toward the root) and downward edges (toward the leaves) respectively. I made sure these were recoverable by having all of the upward edges come before the downward edges in the edge_index. These are here for the purpose of determining which nodes are children and which are parents, in the forward() method of the MPNN.

I've run the training loop with SGD, with LR=0.01 and momentum=0.8.

Any help/advice would be so appreciated, I've been racking my brain on this for months :(

**Comment by Coprosmo** [comment]
I've just checked the gradients for the different parts of the MPNN, and it looks like the gradients for the aggregation MLP (MLP_aggr in code) are consistently an order of magnitude higher than those for the message-encoding MLPs (MLP_edge and MLP_edge_hat).

**Comment by rusty1s** [comment]
Those are definitely interesting findings. One reason for gradients in the message MLP being lower is because of mean aggregation, which diminish the output gradient by a factor dependent on the number of neighbors: That is, in case of many neighbors, each message will only get trained by a small margin. My guess is that you will see "better" gradients when using "max" or "add" aggregation.

I do not have any more advice for you (the gradients are definitely correct). I feel that is an interesting problem and as far as I know, little research has been investigated on what message functions actually learn.

**Comment by Coprosmo** [comment]
Good thinking! Thanks for the advice, I'll test out the other aggregation functions :)
